// 0x.js/lib/test/utils/fill_scenarios.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("@0xproject/utils");
var src_1 = require("../../src");
var artifacts_1 = require("../../src/artifacts");
var dummy_token_1 = require("../../src/contract_wrappers/generated/dummy_token");
var order_factory_1 = require("../utils/order_factory");
var constants_1 = require("./constants");
var INITIAL_COINBASE_TOKEN_SUPPLY_IN_UNITS = new utils_1.BigNumber(100);
var FillScenarios = /** @class */ (function () {
    function FillScenarios(zeroEx, userAddresses, tokens, zrxTokenAddress, exchangeContractAddress) {
        this._zeroEx = zeroEx;
        this._userAddresses = userAddresses;
        this._tokens = tokens;
        this._coinbase = userAddresses[0];
        this._zrxTokenAddress = zrxTokenAddress;
        this._exchangeContractAddress = exchangeContractAddress;
    }
    FillScenarios.prototype.initTokenBalancesAsync = function () {
        return __awaiter(this, void 0, void 0, function () {
            var web3Wrapper, _i, _a, token, contractInstance, defaults, dummyToken, tokenSupply, txHash;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        web3Wrapper = this._zeroEx._web3Wrapper;
                        _i = 0, _a = this._tokens;
                        _b.label = 1;
                    case 1:
                        if (!(_i < _a.length)) return [3 /*break*/, 5];
                        token = _a[_i];
                        if (!(token.symbol !== 'ZRX' && token.symbol !== 'WETH')) return [3 /*break*/, 4];
                        contractInstance = web3Wrapper.getContractInstance(artifacts_1.artifacts.DummyTokenArtifact.abi, token.address);
                        defaults = {};
                        dummyToken = new dummy_token_1.DummyTokenContract(contractInstance, defaults);
                        tokenSupply = src_1.ZeroEx.toBaseUnitAmount(INITIAL_COINBASE_TOKEN_SUPPLY_IN_UNITS, token.decimals);
                        return [4 /*yield*/, dummyToken.setBalance.sendTransactionAsync(this._coinbase, tokenSupply, {
                                from: this._coinbase,
                            })];
                    case 2:
                        txHash = _b.sent();
                        return [4 /*yield*/, this._zeroEx.awaitTransactionMinedAsync(txHash)];
                    case 3:
                        _b.sent();
                        _b.label = 4;
                    case 4:
                        _i++;
                        return [3 /*break*/, 1];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    FillScenarios.prototype.createFillableSignedOrderAsync = function (makerTokenAddress, takerTokenAddress, makerAddress, takerAddress, fillableAmount, expirationUnixTimestampSec) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.createAsymmetricFillableSignedOrderAsync(makerTokenAddress, takerTokenAddress, makerAddress, takerAddress, fillableAmount, fillableAmount, expirationUnixTimestampSec)];
            });
        });
    };
    FillScenarios.prototype.createFillableSignedOrderWithFeesAsync = function (makerTokenAddress, takerTokenAddress, makerFee, takerFee, makerAddress, takerAddress, fillableAmount, feeRecepient, expirationUnixTimestampSec) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this._createAsymmetricFillableSignedOrderWithFeesAsync(makerTokenAddress, takerTokenAddress, makerFee, takerFee, makerAddress, takerAddress, fillableAmount, fillableAmount, feeRecepient, expirationUnixTimestampSec)];
            });
        });
    };
    FillScenarios.prototype.createAsymmetricFillableSignedOrderAsync = function (makerTokenAddress, takerTokenAddress, makerAddress, takerAddress, makerFillableAmount, takerFillableAmount, expirationUnixTimestampSec) {
        return __awaiter(this, void 0, void 0, function () {
            var makerFee, takerFee, feeRecepient;
            return __generator(this, function (_a) {
                makerFee = new utils_1.BigNumber(0);
                takerFee = new utils_1.BigNumber(0);
                feeRecepient = constants_1.constants.NULL_ADDRESS;
                return [2 /*return*/, this._createAsymmetricFillableSignedOrderWithFeesAsync(makerTokenAddress, takerTokenAddress, makerFee, takerFee, makerAddress, takerAddress, makerFillableAmount, takerFillableAmount, feeRecepient, expirationUnixTimestampSec)];
            });
        });
    };
    FillScenarios.prototype.createPartiallyFilledSignedOrderAsync = function (makerTokenAddress, takerTokenAddress, takerAddress, fillableAmount, partialFillAmount) {
        return __awaiter(this, void 0, void 0, function () {
            var makerAddress, signedOrder, shouldThrowOnInsufficientBalanceOrAllowance;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        makerAddress = this._userAddresses[0];
                        return [4 /*yield*/, this.createAsymmetricFillableSignedOrderAsync(makerTokenAddress, takerTokenAddress, makerAddress, takerAddress, fillableAmount, fillableAmount)];
                    case 1:
                        signedOrder = _a.sent();
                        shouldThrowOnInsufficientBalanceOrAllowance = false;
                        return [4 /*yield*/, this._zeroEx.exchange.fillOrderAsync(signedOrder, partialFillAmount, shouldThrowOnInsufficientBalanceOrAllowance, takerAddress)];
                    case 2:
                        _a.sent();
                        return [2 /*return*/, signedOrder];
                }
            });
        });
    };
    FillScenarios.prototype._createAsymmetricFillableSignedOrderWithFeesAsync = function (makerTokenAddress, takerTokenAddress, makerFee, takerFee, makerAddress, takerAddress, makerFillableAmount, takerFillableAmount, feeRecepient, expirationUnixTimestampSec) {
        return __awaiter(this, void 0, void 0, function () {
            var signedOrder;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, Promise.all([
                            this._increaseBalanceAndAllowanceAsync(makerTokenAddress, makerAddress, makerFillableAmount),
                            this._increaseBalanceAndAllowanceAsync(takerTokenAddress, takerAddress, takerFillableAmount),
                        ])];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, Promise.all([
                                this._increaseBalanceAndAllowanceAsync(this._zrxTokenAddress, makerAddress, makerFee),
                                this._increaseBalanceAndAllowanceAsync(this._zrxTokenAddress, takerAddress, takerFee),
                            ])];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, order_factory_1.orderFactory.createSignedOrderAsync(this._zeroEx, makerAddress, takerAddress, makerFee, takerFee, makerFillableAmount, makerTokenAddress, takerFillableAmount, takerTokenAddress, this._exchangeContractAddress, feeRecepient, expirationUnixTimestampSec)];
                    case 3:
                        signedOrder = _a.sent();
                        return [2 /*return*/, signedOrder];
                }
            });
        });
    };
    FillScenarios.prototype._increaseBalanceAndAllowanceAsync = function (tokenAddress, address, amount) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (amount.isZero() || address === src_1.ZeroEx.NULL_ADDRESS) {
                            return [2 /*return*/]; // noop
                        }
                        return [4 /*yield*/, Promise.all([
                                this._increaseBalanceAsync(tokenAddress, address, amount),
                                this._increaseAllowanceAsync(tokenAddress, address, amount),
                            ])];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    FillScenarios.prototype._increaseBalanceAsync = function (tokenAddress, address, amount) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this._zeroEx.token.transferAsync(tokenAddress, this._coinbase, address, amount)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    FillScenarios.prototype._increaseAllowanceAsync = function (tokenAddress, address, amount) {
        return __awaiter(this, void 0, void 0, function () {
            var oldMakerAllowance, newMakerAllowance;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this._zeroEx.token.getProxyAllowanceAsync(tokenAddress, address)];
                    case 1:
                        oldMakerAllowance = _a.sent();
                        newMakerAllowance = oldMakerAllowance.plus(amount);
                        return [4 /*yield*/, this._zeroEx.token.setProxyAllowanceAsync(tokenAddress, address, newMakerAllowance)];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    return FillScenarios;
}());
exports.FillScenarios = FillScenarios;
