// 0x.js/lib/test/remaining_fillable_calculator_test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var _this = this;
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("@0xproject/utils");
var chai = require("chai");
require("mocha");
var _0x_1 = require("../src/0x");
var remaining_fillable_calculator_1 = require("../src/order_watcher/remaining_fillable_calculator");
var chai_setup_1 = require("./utils/chai_setup");
chai_setup_1.chaiSetup.configure();
var expect = chai.expect;
describe('RemainingFillableCalculator', function () {
    var calculator;
    var signedOrder;
    var transferrableMakerTokenAmount;
    var transferrableMakerFeeTokenAmount;
    var remainingMakerTokenAmount;
    var makerAmount;
    var takerAmount;
    var makerFeeAmount;
    var isMakerTokenZRX;
    var makerToken = '0x1';
    var takerToken = '0x2';
    var decimals = 4;
    var zero = new utils_1.BigNumber(0);
    var zeroAddress = '0x0';
    var signature = { v: 27, r: '', s: '' };
    beforeEach(function () { return __awaiter(_this, void 0, void 0, function () {
        var _a, _b;
        return __generator(this, function (_c) {
            _a = [
                _0x_1.ZeroEx.toBaseUnitAmount(new utils_1.BigNumber(50), decimals),
                _0x_1.ZeroEx.toBaseUnitAmount(new utils_1.BigNumber(5), decimals),
                _0x_1.ZeroEx.toBaseUnitAmount(new utils_1.BigNumber(1), decimals),
            ], makerAmount = _a[0], takerAmount = _a[1], makerFeeAmount = _a[2];
            _b = [
                _0x_1.ZeroEx.toBaseUnitAmount(new utils_1.BigNumber(50), decimals),
                _0x_1.ZeroEx.toBaseUnitAmount(new utils_1.BigNumber(5), decimals),
            ], transferrableMakerTokenAmount = _b[0], transferrableMakerFeeTokenAmount = _b[1];
            return [2 /*return*/];
        });
    }); });
    function buildSignedOrder() {
        return {
            ecSignature: signature,
            exchangeContractAddress: zeroAddress,
            feeRecipient: zeroAddress,
            maker: zeroAddress,
            taker: zeroAddress,
            makerFee: makerFeeAmount,
            takerFee: zero,
            makerTokenAmount: makerAmount,
            takerTokenAmount: takerAmount,
            makerTokenAddress: makerToken,
            takerTokenAddress: takerToken,
            salt: zero,
            expirationUnixTimestampSec: zero,
        };
    }
    describe('Maker token is NOT ZRX', function () {
        before(function () { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                isMakerTokenZRX = false;
                return [2 /*return*/];
            });
        }); });
        it('calculates the correct amount when unfilled and funds available', function () {
            signedOrder = buildSignedOrder();
            remainingMakerTokenAmount = signedOrder.makerTokenAmount;
            calculator = new remaining_fillable_calculator_1.RemainingFillableCalculator(signedOrder, isMakerTokenZRX, transferrableMakerTokenAmount, transferrableMakerFeeTokenAmount, remainingMakerTokenAmount);
            expect(calculator.computeRemainingMakerFillable()).to.be.bignumber.equal(remainingMakerTokenAmount);
        });
        it('calculates the correct amount when partially filled and funds available', function () {
            signedOrder = buildSignedOrder();
            remainingMakerTokenAmount = _0x_1.ZeroEx.toBaseUnitAmount(new utils_1.BigNumber(1), decimals);
            calculator = new remaining_fillable_calculator_1.RemainingFillableCalculator(signedOrder, isMakerTokenZRX, transferrableMakerTokenAmount, transferrableMakerFeeTokenAmount, remainingMakerTokenAmount);
            expect(calculator.computeRemainingMakerFillable()).to.be.bignumber.equal(remainingMakerTokenAmount);
        });
        it('calculates the amount to be 0 when all fee funds are transferred', function () {
            signedOrder = buildSignedOrder();
            transferrableMakerFeeTokenAmount = zero;
            remainingMakerTokenAmount = signedOrder.makerTokenAmount;
            calculator = new remaining_fillable_calculator_1.RemainingFillableCalculator(signedOrder, isMakerTokenZRX, transferrableMakerTokenAmount, transferrableMakerFeeTokenAmount, remainingMakerTokenAmount);
            expect(calculator.computeRemainingMakerFillable()).to.be.bignumber.equal(zero);
        });
        it('calculates the correct amount when balance is less than remaining fillable', function () {
            signedOrder = buildSignedOrder();
            var partiallyFilledAmount = _0x_1.ZeroEx.toBaseUnitAmount(new utils_1.BigNumber(2), decimals);
            remainingMakerTokenAmount = signedOrder.makerTokenAmount.minus(partiallyFilledAmount);
            transferrableMakerTokenAmount = remainingMakerTokenAmount.minus(partiallyFilledAmount);
            calculator = new remaining_fillable_calculator_1.RemainingFillableCalculator(signedOrder, isMakerTokenZRX, transferrableMakerTokenAmount, transferrableMakerFeeTokenAmount, remainingMakerTokenAmount);
            expect(calculator.computeRemainingMakerFillable()).to.be.bignumber.equal(transferrableMakerTokenAmount);
        });
        describe('Order to Fee Ratio is < 1', function () {
            beforeEach(function () { return __awaiter(_this, void 0, void 0, function () {
                var _a;
                return __generator(this, function (_b) {
                    _a = [
                        _0x_1.ZeroEx.toBaseUnitAmount(new utils_1.BigNumber(3), decimals),
                        _0x_1.ZeroEx.toBaseUnitAmount(new utils_1.BigNumber(6), decimals),
                        _0x_1.ZeroEx.toBaseUnitAmount(new utils_1.BigNumber(6), decimals),
                    ], makerAmount = _a[0], takerAmount = _a[1], makerFeeAmount = _a[2];
                    return [2 /*return*/];
                });
            }); });
            it('calculates the correct amount when funds unavailable', function () {
                signedOrder = buildSignedOrder();
                remainingMakerTokenAmount = signedOrder.makerTokenAmount;
                var transferredAmount = _0x_1.ZeroEx.toBaseUnitAmount(new utils_1.BigNumber(2), decimals);
                transferrableMakerTokenAmount = remainingMakerTokenAmount.minus(transferredAmount);
                calculator = new remaining_fillable_calculator_1.RemainingFillableCalculator(signedOrder, isMakerTokenZRX, transferrableMakerTokenAmount, transferrableMakerFeeTokenAmount, remainingMakerTokenAmount);
                expect(calculator.computeRemainingMakerFillable()).to.be.bignumber.equal(transferrableMakerTokenAmount);
            });
        });
        describe('Ratio is not evenly divisble', function () {
            beforeEach(function () { return __awaiter(_this, void 0, void 0, function () {
                var _a;
                return __generator(this, function (_b) {
                    _a = [
                        _0x_1.ZeroEx.toBaseUnitAmount(new utils_1.BigNumber(3), decimals),
                        _0x_1.ZeroEx.toBaseUnitAmount(new utils_1.BigNumber(7), decimals),
                        _0x_1.ZeroEx.toBaseUnitAmount(new utils_1.BigNumber(7), decimals),
                    ], makerAmount = _a[0], takerAmount = _a[1], makerFeeAmount = _a[2];
                    return [2 /*return*/];
                });
            }); });
            it('calculates the correct amount when funds unavailable', function () {
                signedOrder = buildSignedOrder();
                remainingMakerTokenAmount = signedOrder.makerTokenAmount;
                var transferredAmount = _0x_1.ZeroEx.toBaseUnitAmount(new utils_1.BigNumber(2), decimals);
                transferrableMakerTokenAmount = remainingMakerTokenAmount.minus(transferredAmount);
                calculator = new remaining_fillable_calculator_1.RemainingFillableCalculator(signedOrder, isMakerTokenZRX, transferrableMakerTokenAmount, transferrableMakerFeeTokenAmount, remainingMakerTokenAmount);
                var calculatedFillableAmount = calculator.computeRemainingMakerFillable();
                expect(calculatedFillableAmount.lessThanOrEqualTo(transferrableMakerTokenAmount)).to.be.true();
                expect(calculatedFillableAmount).to.be.bignumber.greaterThan(new utils_1.BigNumber(0));
                var orderToFeeRatio = signedOrder.makerTokenAmount.dividedBy(signedOrder.makerFee);
                var calculatedFeeAmount = calculatedFillableAmount.dividedBy(orderToFeeRatio);
                expect(calculatedFeeAmount).to.be.bignumber.lessThan(transferrableMakerFeeTokenAmount);
            });
        });
    });
    describe('Maker Token is ZRX', function () {
        before(function () { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                isMakerTokenZRX = true;
                return [2 /*return*/];
            });
        }); });
        it('calculates the correct amount when unfilled and funds available', function () {
            signedOrder = buildSignedOrder();
            transferrableMakerTokenAmount = makerAmount.plus(makerFeeAmount);
            transferrableMakerFeeTokenAmount = transferrableMakerTokenAmount;
            remainingMakerTokenAmount = signedOrder.makerTokenAmount;
            calculator = new remaining_fillable_calculator_1.RemainingFillableCalculator(signedOrder, isMakerTokenZRX, transferrableMakerTokenAmount, transferrableMakerFeeTokenAmount, remainingMakerTokenAmount);
            expect(calculator.computeRemainingMakerFillable()).to.be.bignumber.equal(remainingMakerTokenAmount);
        });
        it('calculates the correct amount when partially filled and funds available', function () {
            signedOrder = buildSignedOrder();
            remainingMakerTokenAmount = _0x_1.ZeroEx.toBaseUnitAmount(new utils_1.BigNumber(1), decimals);
            calculator = new remaining_fillable_calculator_1.RemainingFillableCalculator(signedOrder, isMakerTokenZRX, transferrableMakerTokenAmount, transferrableMakerFeeTokenAmount, remainingMakerTokenAmount);
            expect(calculator.computeRemainingMakerFillable()).to.be.bignumber.equal(remainingMakerTokenAmount);
        });
        it('calculates the amount to be 0 when all fee funds are transferred', function () {
            signedOrder = buildSignedOrder();
            transferrableMakerTokenAmount = zero;
            transferrableMakerFeeTokenAmount = zero;
            remainingMakerTokenAmount = signedOrder.makerTokenAmount;
            calculator = new remaining_fillable_calculator_1.RemainingFillableCalculator(signedOrder, isMakerTokenZRX, transferrableMakerTokenAmount, transferrableMakerFeeTokenAmount, remainingMakerTokenAmount);
            expect(calculator.computeRemainingMakerFillable()).to.be.bignumber.equal(zero);
        });
        it('calculates the correct amount when balance is less than remaining fillable', function () {
            signedOrder = buildSignedOrder();
            var partiallyFilledAmount = _0x_1.ZeroEx.toBaseUnitAmount(new utils_1.BigNumber(2), decimals);
            remainingMakerTokenAmount = signedOrder.makerTokenAmount.minus(partiallyFilledAmount);
            transferrableMakerTokenAmount = remainingMakerTokenAmount.minus(partiallyFilledAmount);
            transferrableMakerFeeTokenAmount = transferrableMakerTokenAmount;
            var orderToFeeRatio = signedOrder.makerTokenAmount.dividedToIntegerBy(signedOrder.makerFee);
            var expectedFillableAmount = new utils_1.BigNumber(450980);
            calculator = new remaining_fillable_calculator_1.RemainingFillableCalculator(signedOrder, isMakerTokenZRX, transferrableMakerTokenAmount, transferrableMakerFeeTokenAmount, remainingMakerTokenAmount);
            var calculatedFillableAmount = calculator.computeRemainingMakerFillable();
            var numberOfFillsInRatio = calculatedFillableAmount.dividedToIntegerBy(orderToFeeRatio);
            var calculatedFillableAmountPlusFees = calculatedFillableAmount.plus(numberOfFillsInRatio);
            expect(calculatedFillableAmountPlusFees).to.be.bignumber.lessThan(transferrableMakerTokenAmount);
            expect(calculatedFillableAmountPlusFees).to.be.bignumber.lessThan(remainingMakerTokenAmount);
            expect(calculatedFillableAmount).to.be.bignumber.equal(expectedFillableAmount);
            expect(numberOfFillsInRatio.decimalPlaces()).to.be.equal(0);
        });
    });
});
